l10n_it_fatturapa/models/account.py

```diff
@@ -158,15 +137,15 @@ class DiscountRisePrice(models.Model):
 
     name = fields.Selection([("SC", "Discount"), ("MG", "Supplement")], "Type")
     percentage = fields.Float("Percentage")
-    amount = fields.Float("Amount", digits=dp.get_precision("Discount"))
+    amount = fields.Float("Amount", digits="Discount")
     invoice_line_id = fields.Many2one(
```

Comment:
Ma a livello di migrazione dati, l'ID contenuto nella colonna `invoice_line_id` a quale `account.move.line` punterà sulla 14?
Voglio dire: se nella 12 il campo contiene l'ID di una `account.invoice.line`, nella 14 quello stesso ID potrebbe non esistere o puntare a un'altra riga di `account.move.line`.
Giusto?

Reply:
Immagino servirà qualcosa di analogo a https://github.com/OCA/OpenUpgrade/pull/2387/files

Reply:
È un problema a monte, se account è stato migrato, allora le vecchie righe di invoice, che erano sparse su + tabelle, sono state già migrate tutte in account_move_line, con i giusti campi, e tutte le relazioni sono state aggiornate. L'ID è quasi certamente cambiato ma noi quello nuovo non lo sappiamo. Deve essere lo script di migrazione di account che deve cercare tutte le relazioni che fanno riferimento a account_invoice_line (vecchia tabella) e aggiornarle con il nuovi ID.

La cosa in pratica deve essere già successa. Siccome il modulo account migra prima di noi, la tabella account_invoice_line non esiste già più quando viene invocato il nostro script, o sbaglio?

Reply:
NOTA: va provato in questo modo. Installare purchase_stock e l10n_it_fatturapa e poi lanciare tutti i test. Ce ne sono alcuni che falliscono ancora.

Reply:
Sì, se sulla 14 la tabella `account_invoice_line` e la colonna `old_invoice_line_id` di `account_move_line` non ci sono più, allora bisogna fare lo script sulla 13.
Vedi
https://github.com/OCA/OpenUpgrade/blob/1602ea5a56c225049cccbed8181e18fa0851a9c9/addons/membership/migrations/13.0.1.0/pre-migration.py#L21-L25

Reply:
Ah, ok, adesso ci sono arrivato. 

Sì, serve sicuramente qualcosa di simile, e esattamente nello stesso punto. "chi lo ha fatto deve aver già sistemato i riferimenti delle relazioni esistenti" intendevo proprio quello.

Io vedevo le modifiche relative ai cambiamenti di l10n_it_fatturapa come una fase successiva all'aggiustamento del db dovuto ai cambiamenti del modulo account. Quello conviene farlo come migration 13.0.0.0.

Ma la prassi qual è? Lo mettiamo noi nel nostro modulo o lo si deve aggiungere al repo Openupgrade?

Reply:
Va messo nella cartella migrations del nostro modulo

Reply:
> Quello conviene farlo come migration 13.0.0.0

A meno che i dati non restino presenti, quatti quatti, nei DB 14; in quel caso potremmo mettere lo script di migrazione direttamente nel modulo della 14.

Se qualcuno con Odoo Enterprise vuole provare... Altrimenti dobbiamo aspettare, o fare, gli script di migrazione di openupgrade e testare una migrazione.

Reply:
Nel DB migrato da enterprise su cui sto provando, la tabella `account_invoice` c'è.
Per recuperare il vecchio numero fattura, se il DB è migrato da OpenUpgrade allora ci sarà una colonna a DB `old_invoice_id` nella tabella `account_move` e analogamente c'è `old_invoice_line_id` in `account_move_line`.

Secondo me va comunque pensato un modo approccio più generico per supportare anche i DB aggiornati con Enterprise.

Per recuperare la fattura collegata puoi fare come ho proposto in https://github.com/Borruso/l10n-italy/pull/5, che è lo stesso modo utilizzato per valorizzare `old_invoice_id` in OpenUpgrade: https://github.com/OCA/OpenUpgrade/blob/d8c688d4419a94da9dee46ec90724033cbd262b5/addons/account/migrations/13.0.1.1/pre-migration.py#L237-L244

Per recuperare la vecchia riga fattura puoi vedere come OpenUpgrade si calcola `old_invoice_line_id`:
https://github.com/OCA/OpenUpgrade/blob/d8c688d4419a94da9dee46ec90724033cbd262b5/addons/account/migrations/13.0.1.1/post-migration.py#L221-L226


Per quanto riguarda i test di tutti i moduli, una curiosità: hai provato se succede anche in v12?

Reply:
Ho aggiunto lo script che dovrebbe fare la cosa giusta... sulla carta. Non so come provarlo. Tocca fare l'intero giro openupgrade di un db demo della 12, mi sa.

Reply:
Come non detto. Ho letto dopo il commento. Adesso vedo di seguire il secondo approccio (adesso supporta solo openupgrade).

Reply:
> Per recuperare la fattura collegata puoi fare come ho proposto in [Borruso#5](https://github.com/Borruso/l10n-italy/pull/5), che è lo stesso modo utilizzato per valorizzare `old_invoice_id` in OpenUpgrade: https://github.com/OCA/OpenUpgrade/blob/d8c688d4419a94da9dee46ec90724033cbd262b5/addons/account/migrations/13.0.1.1/pre-migration.py#L237-L244
> 
> Per recuperare la vecchia riga fattura puoi vedere come OpenUpgrade si calcola `old_invoice_line_id`:
> https://github.com/OCA/OpenUpgrade/blob/d8c688d4419a94da9dee46ec90724033cbd262b5/addons/account/migrations/13.0.1.1/post-migration.py#L221-L226


Primo dubbio: siamo sicuri che upgrade di EE faccia la stessa cosa? Ossia preservi il vecchio move_id?

Mi spiego meglio.

Possiamo provare a "calcolare" noi old_invoice_id nel caso di EE solo se usano lo stesso metodo di openupgrade. E in più, se vedi il codice che hai linkato, ci sono due fasi: la prima va in update, ossia copia i dati dalle vecchie fatture alla move _corrispondenti_ preservando account_invoice.move_id = account_move.id. Il problema è la seconda, che va in insert: se non c'è il campo old_invoice_id non abbiamo modo di conoscere i nuovi id nella account_move corrispondenti alle righe che prima non c'erano (fatture in 'draft' o 'cancelled'). 

La stessa cosa succede per le righe. Quelle che già ci sono, vengono aggiornate. E potremmo applicare la stessa logica di abbinamento tra le vecchie e le nuove righe per ricavarne gli id. Ma quelle che non c'erano vengono create ex-novo e i loro id non li possiamo calcolare.

E ora che ci penso, manca la parte che copia i _nostri_ dati.

> Per quanto riguarda i test di tutti i moduli, una curiosità: hai provato se succede anche in v12?

In effetti, no.

Reply:
Secondo me `account_invoice.move_id` ci può aiutare a collegare account_move e account_invoice, questo credo abbia senso indipendentemente da chi ha fatto la migrazione del core.
Anche perché se non ci fossero state le move corrispondenti, la migrazione che ho fatto con lo script in https://github.com/Borruso/l10n-italy/pull/5 sarebbe esplosa.

Per quanto riguarda la righe, secondo me ha senso riportare dati solo nelle move line che prima erano delle invoice line quindi non mi preoccuperei troppo delle nuove righe create o di quelle che rappresentano le imposte o altro.
Le varie condizioni controllate in OpenUpgrade in https://github.com/OCA/OpenUpgrade/blob/d8c688d4419a94da9dee46ec90724033cbd262b5/addons/account/migrations/13.0.1.1/post-migration.py#L221-L240 mi sembra possano darci un buon aiuto per trovare il collegamento.

Per quanto riguarda le INSERT, ho controllato quelle con `INSERT INTO account_move*` e mi pare servano solo a ripopolare la tabella: come campo `old_invoice_*` inseriscono l'ID del vecchio record; ti riferivi a qualche insert in particolare?
Ce ne sono alcune che non considererei, ad esempio vengono create delle move line per le righe delle imposte valorizzando `old_invoice_tax_id`, ma come ho scritto sopra mi preoccuperei solo delle move line che prima erano delle invoice line.

Ho comunque aperto un ticket a Odoo per chiedere delucidazioni su come collegare account_invoice[_line] con account_move[_line] perché magari hanno pure loro inziccato qualcosa nel DB che ci potrebbe aiutare e non l'ho visto.

Reply:
> Secondo me account_invoice.move_id ci può aiutare a collegare account_move e account_invoice, questo credo abbia senso indipendentemente da chi ha fatto la migrazione del core.

Ci aiuta sicuramente, quando non è vuota. Per le fatture in draft, è vuota. Le fatture non in draft non sono un problema.

Quando una account_invoice è in draft, post migrazione ci si trova una account_move completamente nuova, non correlata (e le relative _line ovviamente).